## 실습 1 . 데이터타입에 대한 이해 
2**15 = 32768 의 각 자리수를 더하면 3+2+7+6+8=26 입니다. 

2**1000의 각 자리수를 모두 더하면 얼마입니까?

In [18]:
data1=2**15
print(data1)
type(data1)

32768


int

In [19]:
data2 = 2**1000
print(data2)
type(data2)

10715086071862673209484250490600018105614048117055336074437503883703510511249361224931983788156958581275946729175531468251871452856923140435984577574698574803934567774824230985421074605062371141877954182153046474983581941267398767559165543946077062914571196477686542167660429831652624386837205668069376


int

In [37]:
def add_number(da1,da2):
    result = da1 + da2
    return result


In [38]:
d2 = str(data2) 
type(d2)
d1= 0
for i in range(len(d2)):
    character = d2[i]
    character1 = int(character) 
    print(character)


1
0
7
1
5
0
8
6
0
7
1
8
6
2
6
7
3
2
0
9
4
8
4
2
5
0
4
9
0
6
0
0
0
1
8
1
0
5
6
1
4
0
4
8
1
1
7
0
5
5
3
3
6
0
7
4
4
3
7
5
0
3
8
8
3
7
0
3
5
1
0
5
1
1
2
4
9
3
6
1
2
2
4
9
3
1
9
8
3
7
8
8
1
5
6
9
5
8
5
8
1
2
7
5
9
4
6
7
2
9
1
7
5
5
3
1
4
6
8
2
5
1
8
7
1
4
5
2
8
5
6
9
2
3
1
4
0
4
3
5
9
8
4
5
7
7
5
7
4
6
9
8
5
7
4
8
0
3
9
3
4
5
6
7
7
7
4
8
2
4
2
3
0
9
8
5
4
2
1
0
7
4
6
0
5
0
6
2
3
7
1
1
4
1
8
7
7
9
5
4
1
8
2
1
5
3
0
4
6
4
7
4
9
8
3
5
8
1
9
4
1
2
6
7
3
9
8
7
6
7
5
5
9
1
6
5
5
4
3
9
4
6
0
7
7
0
6
2
9
1
4
5
7
1
1
9
6
4
7
7
6
8
6
5
4
2
1
6
7
6
6
0
4
2
9
8
3
1
6
5
2
6
2
4
3
8
6
8
3
7
2
0
5
6
6
8
0
6
9
3
7
6


In [42]:
def add_number(da1,da2):
    result = da1 + da2
    return result

data2 = 2**1000
data2 = str(data2)
result =0

for i in range(len(d2)):
    character = d2[i]
    character1 = int(character) 
    result = add_number(character1,result)
print(result)


1366


### 답 ) 1366

#  실습 2 . 로또번호를 주문하는 클래스(class lotto)를 생성하시오.
- 로또는 1~10까지 주문가능 
- 6개의 숫자로 된 로또는 오름차순으로 정렬 
- 주문한 개수만큼의 일련번호를달아 로또번호를 출력 
- OOP 기반으로 클래스에서 메소드를 최대한 모듈화 

In [44]:
#0~45까지 숫자 중 임의로 6개 뽑는 로또 

import random 
random.sample(list(range(1,46)),6)

[29, 8, 31, 44, 5, 19]

In [118]:
def lotto():
    i = input("로또를 몇번 뽑을까요?")
    i=int(i)
    import random 
    if 0<i<11:
        for idx in range(i) :
            idx=random.sample(list(range(1,46)),6)
            print(idx)

In [120]:
lotto()

로또를 몇번 뽑을까요?5
[38, 3, 32, 2, 8, 42]
[24, 3, 25, 44, 10, 16]
[5, 12, 24, 10, 14, 36]
[26, 28, 16, 39, 45, 38]
[32, 16, 44, 31, 28, 19]


In [165]:
def lotto(i):
    import random 
    if 0<i<11:
        for idx in range(i) :
            s=random.sample(list(range(1,46)),6)
            print(s)

In [166]:
lotto(5)

[21, 20, 42, 18, 6, 13]
[12, 3, 39, 40, 36, 26]
[13, 12, 24, 20, 4, 25]
[6, 32, 17, 41, 21, 33]
[5, 29, 28, 16, 13, 14]


In [174]:
#random 없이 만들어보기 
i = input("로또를 몇번 뽑을까요?")
i=int(i)

ball = [x for x in range(1, 46)] 
for x in range(i):
    random.shuffle(ball)     # 섞기
    ball_selected = ball[:6]  # 6개만 선택
    ball_selected.sort()     # 선택된 번호를 정렬
    print(ball_selected)     # 출력


로또를 몇번 뽑을까요?5
[7, 8, 13, 16, 22, 32]
[2, 4, 5, 10, 25, 41]
[12, 14, 27, 32, 37, 43]
[1, 19, 20, 32, 33, 35]
[7, 12, 29, 32, 40, 43]


In [178]:
#일련번호 달기
i = input("로또를 몇번 뽑을까요?")
i=int(i)

ball = [x for x in range(1, 46)] 
for x in range(i):
    random.shuffle(ball)     # 섞기
    ball_selected = ball[:6]  # 6개만 선택
    ball_selected.sort()     # 선택된 번호를 정렬
    print("{} : {}". format((x+1),(ball_selected)))     # 출력

로또를 몇번 뽑을까요?5
1 : [9, 13, 14, 18, 33, 39]
2 : [1, 7, 24, 31, 34, 39]
3 : [4, 7, 11, 28, 38, 42]
4 : [3, 8, 23, 24, 35, 43]
5 : [1, 4, 12, 19, 31, 42]


# 실습3 . 특정 기념일로부터 얼마나 경과되었는지를 알려주는 함수를 만드시오 
- 경과일자가 1000일이 넘는 경우, 천단위로 (,)를 붙여서 출력하기 
- 구현할 함수의 옵션 : 
   from datetime import datetime 
   def gerMemorialDay(year,month,day,mem_day='기념일',is_msg=True):
       return remain_year,remain_month, remain_day
- 출력형태
    getMemorialDay(2014,4,16,'세월호침몰사고일')
    
    오늘은 "세월호침몰사고일"로부터 1,658일 경과되었고, 달 수로는 55개월째, 연수로는 4년째 되었습니다!
    
    _y,_m,elapsed_day = getMemorialDay(2018,12,25,'크리스마스',False)
    
    크리스마스까지는 %d일 남았습니다! % -elapsed_day

In [54]:
#날짜세는 함수 만들기 

from datetime import datetime

def getMemorialDay(year,month,day,mem_day='기념일',is_msg=True):
    mem_day = datetime(year,month,day)
    today = datetime.today()
    elapsed = today - mem_day
    elapsed_day = int(elapsed.days)
    elapsed_month = int(elapsed_day/30)
    elapsed_year = int(elapsed_day/365)
    print(elapsed_day, elapsed_month, elapsed_year)
    

In [55]:
getMemorialDay(2014,4,16, '세월호침몰사고일')

1659 55 4


In [63]:
#콤마찍는 함수만들기 
def comma(data):
    data = str(data)
    r_data= data[::-1]
    print(r_data)  #우선 뒤집어준다 -> 뒤집어서 세번째 자리수면 ","을 추가해줌
    
    
comma(123456789)

987654321


In [80]:
#콤마찍는 함수만들기 
def comma(data):
    data = str(data)
    r_data= data[::-1]
    
    date=""
    for i in range(len(r_data)): 
        char= r_data[i]
        com=","
        
        if i%3==0 and i>0 :
            date += com
            
        date += char
        
    return date[::-1]
    
comma(123456789)

'123,456,789'

In [103]:
#합치기 
def comma(data):
    data = str(data)
    r_data= data[::-1]
    
    date=""
    for i in range(len(r_data)): 
        char= r_data[i]
        com=","
        
        if i%3==0 and i>0 :
            date += com
            
        date += char
        
    return date[::-1]

from datetime import datetime

def getMemorialDay(year,month,day,mem_day='기념일',is_msg=True):
    mem= datetime(year,month,day)
    today = datetime.today()
    
    
    if is_msg: 
        elapsed = today - mem
        elapsed_day = int(elapsed.days)
        elapsed_month = int(elapsed_day/30)
        elapsed_year = int(elapsed_day/365)
        elapsed_day = comma(elapsed_day)      #comma 함수 적용
        print("오늘은 {}로부터 {}일 경과되었고, 달 수로는 {}개월째, 연 수로는 {}년째 되었습니다!!".
                  format(mem_day,elapsed_day,elapsed_month,elapsed_year))
    else: 
        elapsed = mem-today
        elapsed_day = int(elapsed.days)
        elapsed_month = int(elapsed_day/30)
        elapsed_year = int(elapsed_day/365)
        elapsed_day = comma(elapsed_day)
        print("{}까지는 {}일 남았습니다!".format(mem_day,elapsed_day))
        
    


In [104]:
getMemorialDay(2014,4,16, '세월호침몰사고일')

오늘은 세월호침몰사고일로부터 1,659일 경과되었고, 달 수로는 55개월째, 연 수로는 4년째 되었습니다!!


오늘은 세월호침몰사고일로부터 1,659일 경과되었고, 달 수로는 55개월째, 연 수로는 4년째 되었습니다!!


In [105]:
getMemorialDay(2018,12,25,'크리스마스',False)

크리스마스까지는 54일 남았습니다!


크리스마스까지는 54일 남았습니다!

# 실습 4. reduce를 사용하여 히스토그램(histogram)함수를 작성하시오 
- 아래의 리스트를 입력받아 히스토그램 그리기 
    입력값 : ["cat", "cat", "cat", "sheep", "sheep", "duck", "duck", "duck", "duck" ]

In [183]:
 data = ["cat", "cat", "cat", "sheep", "sheep", "duck", "duck", "duck", "duck" ]

In [187]:
#key값들은 list를 set한 값들임
s_data = set(data)
print(s_data)

{'cat', 'duck', 'sheep'}


In [238]:
##dic을 만들어줌 
len(data)
dic={"cat":0,"sheep":0,"duck":0}
type(dic)
for key in dic:
    value = dic[key]
    value = int(value)
    print("%s ==> %s" % (key, value))

cat ==> 0
sheep ==> 0
duck ==> 0


In [228]:
###key값을 미리주고 value값을 카운팅하려고 함 
def count(a,b):
    dic={"cat":value for "cat","sheep":value for "sheep","duck":value for "duck"}
    x=0
    y=0
    z=0
    if a=="cat":
        x += 1
    if a=="sheep":
        y += 1
    if a=="duck":
        z += 1

    

In [229]:
#안됨;;
data = ["cat", "cat", "cat", "sheep", "sheep", "duck", "duck", "duck", "duck" ]
reduce(count,data,0)

UnboundLocalError: local variable 'x' referenced before assignment

### 다시 해보기->key값을 기준으로 딕셔너리에 추가해주기 

In [252]:
dic ={}
def count(a,b):
    for i in range(len(data)):
        if a!=b: 
            dic[b] =dic[b] +1
        else :
            dic[a] = dic[a] +1
        return dic

In [253]:
data = ["cat", "cat", "cat", "sheep", "sheep", "duck", "duck", "duck", "duck" ]
reduce(count,data,0)

KeyError: 'cat'

In [261]:
dic ={}
def count(dic,data):
    for dic in data:
        if dic[data]=
        
            
            
            
        else :
            
            
        return dic

NameError: name 'bvalue' is not defined

In [260]:
data = ["cat", "cat", "cat", "sheep", "sheep", "duck", "duck", "duck", "duck" ]
reduce(count,data,0)dic


KeyError: 'cat'

In [278]:
dic ={}
def count(dic,data):
    for i in data:
            dic[i] = 1      
            #key값 추가법
        else :
            dic.keys(i) += 1
            #value 카운트?
                        
        print(dic)

SyntaxError: can't assign to function call (<ipython-input-278-8e170db46c28>, line 7)

## count 함수 만들기

In [300]:
data = ["cat", "cat", "cat", "sheep", "sheep", "duck", "duck", "duck", "duck" ]
dic ={}
def count(dic,data):
    if data in dic.keys():
        dic[data] += 1
    else: 
        dic[data] = 1
    return dic
    

In [301]:
data = ["cat", "cat", "cat", "sheep", "sheep", "duck", "duck", "duck", "duck" ]
reduce(count,data,{})

{'cat': 3, 'sheep': 2, 'duck': 4}

#### 히스토그램 그리기 : input: list => output: histogram dict

In [305]:
result = (reduce(count,data,{}))

In [306]:
print(result)

{'cat': 3, 'sheep': 2, 'duck': 4}


In [319]:
#result의 key값의 갯수만큼 for을 돌릴것 
# 첫번째 key인 고양이는 : value값이 3이다 
#value 1의 값은 = 하나씩에 할당한다. 
#이때 value1은 숫자값  


def draw_histogram(result):
    for i in result:
        num = "=" * int(result[i])
        print ("{}  {}".format(result[i],num))
    

In [320]:
draw_histogram(result)

ValueError: invalid literal for int() with base 10: '==='

In [332]:
print (result)

{'cat': '===', 'sheep': '==', 'duck': '===='}


In [326]:
print(result['cat'])

===


## 히스토그램 완성

In [362]:
data = ["cat", "cat", "cat", "sheep", "sheep", "duck", "duck", "duck", "duck" ]
reduce(count,data,{})
result = (reduce(count,data,{}))

def draw_histogram(result):
    for i in result:
        num = "=" * result[i]
        print (i + "\t" + num)
    return result
        

In [363]:
draw_histogram(result)

cat	===
sheep	==
duck	====


{'cat': 3, 'sheep': 2, 'duck': 4}